pykern/pkresource.py

```diff
@@ -8,34 +8,78 @@
 
 # Root module: Import only builtin packages so avoid dependency issues
 import errno
-import inspect
+import glob
+import importlib
 import os.path
 import pkg_resources
-import re
 
+# TODO(e-carlin): discuss with rn if ok to import pkio
 from pykern import pkinspect
+from pykern import pkio
 from pykern import pksetup
 
 
-def filename(relative_filename, caller_context=None):
+def filename(relative_filename, caller_context=None, packages=None, relpath=False):
     """Return the filename to the resource
 
     Args:
         relative_filename (str): file name relative to package_data directory.
         caller_context (object): Any object from which to get the `root_package`
+        packages (List[str]): Packages to search.
+        relpath (bool): If True path is relative to package package_data dir
 
     Returns:
-        str: absolute path of the resource file
+        str: absolute or relative (to package_data) path of the resource file
     """
-    pkg = pkinspect.root_package(
-        caller_context if caller_context else pkinspect.caller_module())
+    if caller_context and packages:
+        raise ValueError(f'Use only one of caller_context={caller_context} and packages={packages}.')
     assert not os.path.isabs(relative_filename), \
         'must not be an absolute file name={}'.format(relative_filename)
-    fn = os.path.join(pksetup.PACKAGE_DATA, relative_filename)
-    res = pkg_resources.resource_filename(pkg, fn)
-    if not os.path.exists(res):
-        msg = 'resource does not exist for pkg=' + pkg
-        if pkg == '__main__':
-            msg += '; do not call module as a program'
-        raise IOError((errno.ENOENT, msg, res))
+    a = []
+    for f, p in _files(relative_filename, caller_context, packages):
+        a.append(p)
+        if os.path.exists(f):
+            if relpath:
+                f = str(pkio.py_path(
+                    os.path.join(pkg_resources.resource_filename(p, ''), pksetup.PACKAGE_DATA),
+                ).bestrelpath(pkio.py_path(f)))
+            return f
+    msg = f'unable to locate in packages={a}'
+    if '__main__' in a:
+        msg += '; do not call module as a program'
+    raise IOError(errno.ENOENT, msg, relative_filename)
+
+
+def glob_files(relative_path, caller_context=None, packages=None):
+    """Find all paths that match the relative path
+
+    Args:
+        relative_path(str): Path relative to package_data directory.
+        caller_context (object): Any object from which to get the `root_package`.
+        packages (List[str]): Packages to search.
+    Returns:
+        [str]: absolute paths of the matched files
+    """
+    if caller_context and packages:
```

Comment:
This is the same code as `filename()` so share it. If it comes up empty, might be for two reasons: no packages or really not found. glob probably should throw an exception in that case. Shouldn't be called from `__main__` either.

Reply:
a60907c

Reply:
Had to change this to not raise on not found 685c96e